File: src/medium/MaximumTwinSumLinkedList.go
package main

import (
	. "../lib"
	"fmt"
)

func pairSum(head *ListNode) int {
	var tail *ListNode
	n, node := 0, head
	for node != nil {
		tail = &ListNode{Val: node.Val, Next: tail}
		node = node.Next
		n++
	}

	var maxSum int
	for i := 0; i < n/2; i++ {
		if maxSum < head.Val+tail.Val {
			maxSum = head.Val + tail.Val
		}

		head, tail = head.Next, tail.Next
	}

	return maxSum
}

func main() {
	fmt.Println(pairSum(CreateLinkedListFromArray([]int{5, 4, 2, 1})))
	fmt.Println(pairSum(CreateLinkedListFromArray([]int{4, 2, 2, 3})))
	fmt.Println(pairSum(CreateLinkedListFromArray([]int{1, 100000})))
}

File: src/medium/MaximumTwinSumLinkedList.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function pairSum(head: ListNode | null): number {
  let node: ListNode | null = head;
  let lastNodePointer: ListNode | null = head;
  let stack: number[] = [];
  while (lastNodePointer && lastNodePointer.next && node) {
    stack.push(node.val);
    node = node.next;
    lastNodePointer = lastNodePointer.next.next;
  }
  let result: number = 0;
  while (stack && node) {
    result = Math.max(result, stack.pop()! + node.val);
    node = node.next;
  }
  return result;
}
